### ISSUE TITLE
RequestContext.CancellationToken is never used

### ISSUE THREAD

**Issue by heaths** [Azure/azure-sdk-for-net#34988]
For APIs that expose a `RequestContext` instead of a `CancellationToken`, users can only pass a `CancellationToken` by setting `RequestContext.CancellationToken`; however, VS shows 0 references to `RequestContext.CancellationToken` and walking through the pipeline code, everything seems to read `HttpMessage.CancellationToken` which is only set by `HttpPipeline.Send` and `SendAsync` - not something most users are really meant to call themselves.

Looking at generated client code like https://github.com/Azure/azure-sdk-for-net/blob/49cf1a2fb8dc58f166ea4be31efa6d1d0cc63132/sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/Generated/ConversationAnalysisClient.cs#L141-L157 and its request factory like https://github.com/Azure/azure-sdk-for-net/blob/49cf1a2fb8dc58f166ea4be31efa6d1d0cc63132/sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/Generated/ConversationAnalysisClient.cs#L248-L263 I also see nowhere that `RequestContext.CancellationToken` is used; thus, it seems impossible for customers to cancel any methods taking a `RequestContext`.

**Comment by heaths** [comment]
I found this while changing `GetSecretIfExists` using a `CancellationToken` to a `GetSecret` overload taking a `RequestContext` for PR #33767, which is more robust. To work around this, in my custom pipeline (mainly just wrappers) I'm passing `context?.CancellationToken ?? default` to `HttpPipeline.Send` and `SendAsync` but it seems this shouldn't be necessary. Perhaps `HttpPipeline.CreateMessage` should set `HttpMessage.CancellationToken` from the optionally passed-in `RequestContext.CancellationToken`.

**Comment by heaths** [comment]
@annelo-msft pointed out that the `RequestContext.CancellationToken` is getting used via an [extension method](https://github.com/Azure/autorest.csharp/blob/e07e879768792aea9b30c2a8b65abf8f3d596616/src/assets/Generator.Shared/HttpPipelineExtensions.cs#L14). Indeed, I'm seeing this get called in CLU which calls `ProtocolOperationHelpers.ProcessMessage`: https://github.com/Azure/azure-sdk-for-net/blob/49cf1a2fb8dc58f166ea4be31efa6d1d0cc63132/sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/Generated/ConversationAnalysisClient.cs#L176

However, if you call `HttpPipeline.Send` or `SendAsync` directly, this does not appear to be the case. Still seems it would be more robust to have `HttpPipeline.CreateMessage` set the `CancellationToken` as mentioned in my comment above since all code seems to be calling that: https://github.com/Azure/azure-sdk-for-net/blob/49cf1a2fb8dc58f166ea4be31efa6d1d0cc63132/sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/Generated/ConversationAnalysisClient.cs#L250

**Comment by annelo-msft** [comment]
@heaths, given this finding, is there still an issue in CLU?  Or only in KeyVault where the pipeline is being called manually?  What libraries are actually affected?

**Comment by heaths** [comment]
No, it seems there's no issue with CLU. Only when trying to add a convenience method taking a `RequestContext`, which @KrzysztofCwalina, @tg-msft, and @m-nash approved for this; though there's another potential blocker to #33767 outside this problem.

**Comment by annelo-msft** [comment]
Could the extension method be called in the convenience method, or is that not possible?  (If it's not possible, I'm super interested what the blocking reason is.)

**Comment by heaths** [comment]
Sure it could, but now there's an implicit dependency people just have to know to use or behavior may vary (and it's not like many libraries seem to test cancellation). What's wrong with making sure it's always set in a call that everyone has to use anyway like `HttpPipeline.CreateMessage`?

**Comment by annelo-msft** [comment]
Nothing is wrong with it, and I agree it's a nicer approach.  I ask primarily to evaluate whether you're blocked or not, as a means of prioritizing this relative to other tasks.
